lib/cache/resource.js

```diff
@@ -0,0 +1,122 @@
+'use strict';
+
+var clone = require('clone');
+var MemoryCache = require('./memory');
+
+function ResourceCache(cacheId, options) {
+    this._cache = new MemoryCache(cacheId, options);
+}
+
+ResourceCache.prototype.get = function(cacheKey) {
+    return getFromCache.call(this, cacheKey);
+};
+
+function getFromCache(cacheKey) {
+    var entry = clone(this._cache.get(cacheKey));
+
+    // If parent or nested resources not in cache then return undefined.
+    if (typeof entry === 'undefined' || hasExpiredNestedResources.call(this, entry)) {
+        return undefined;
+    }
+
+    getNestedResourcesFromCache.call(this, entry);
+
+    return entry;
+}
+
+function hasExpiredNestedResources(obj) {
+    var keys = Object.keys(obj);
+    for(var i=0; i<keys.length; i++) {
+        var key = keys[i];
+        if (typeof obj[key] !== 'object' || obj[key] == null) {
+            continue;
+        }
+        if (obj[key].hasOwnProperty('key') && obj[key].hasOwnProperty('value') && obj[key].hasOwnProperty('expiration')) {
+            if (typeof getFromCache.call(this, obj[key].key) === 'undefined') {
+                return true;
+            }
+        } else {
+            return hasExpiredNestedResources.call(this, obj[key]);
+        }
+    }
+
+    return false;
+}
+
+function getNestedResourcesFromCache(obj) {
+    var keys = Object.keys(obj);
+    for(var i=0; i<keys.length; i++) {
+        var key = keys[i];
+        if (typeof obj[key] !== 'object' || obj[key] == null) {
+            continue;
+        }
+        if (obj[key].hasOwnProperty('key') && obj[key].hasOwnProperty('value') && obj[key].hasOwnProperty('expiration')) {
+            obj[key] = getFromCache.call(this, obj[key].key);
+        } else {
+            getNestedResourcesFromCache.call(this, obj[key]);
+        }
+    }
+}
+
+ResourceCache.prototype.getAll = function() {
+    return this._cache.getAll();
+};
+
+ResourceCache.prototype.put = function(resource, tags) {
+    var entry = put.call(this, resource, tags);
+    if (entry) {
+        return entry.key;
+    } else {
+        return undefined;
+    }
+};
+
+var put = function(resource, tags) {
+    // Not a resource? Ignore it.
+    if (!resource.hasOwnProperty('$id') || !resource.hasOwnProperty('$expires')) {
+        return;
+    }
+
+    var clonedResource = clone(resource);
```

Comment:
We definitely need it here. The reason is when we insert the original resource, we then update any nested resources within it with the modified cached version of it (i.e. the one with the key, value, and expiration properties that's created when calling memorycache's put method. We need to close this so that we don't update the original resource object.